refactor: ed25519 signature verifier option

kdimak opened the pull request:
closes #1458

Signed-off-by: Dima <dkinoshenko@gmail.com>

Follow-up issue to use Crypto for VC: https://github.com/hyperledger/aries-framework-go/issues/1416
Follow-up issue to use Crypto in JWS/JWT: https://github.com/hyperledger/aries-framework-go/issues/1267

Baha-sk commented on pkg/doc/signature/ed25519signature2018/suite.go:
```diff
@@ -81,18 +93,28 @@ func (s *SignatureSuite) GetDigest(doc []byte) []byte {
+		}
+
+		return nil
+
```
I assume `case *keyset.Handle:` is coming up in another PR?

kdimak replied:
I am basing on a more generic Crypto API interface which takes `kh interface{}`. In this way, not only **tinkrypto** is supported but any other crypto implementation as well. That's my idea.

Baha-sk replied:
I'm ok with generic SignatureSuite API.

Similarly to KMS and Crypto APIs, they offer generic functions. The default implementation is using Tink (`*keyset.Handle`) as argument types.

same here, there should be a case for Tink's *keyset.Handle to call aries-framwork-go's default Crypto.Verify() instead of directly calling de25519 with []byte key

kdimak replied:
It is what is happening in a `default` of the `switch`: we use Crypto instance passed as a functional argument to verify a signature. I think there is no need to introduce a direct dependency on `tinkcrypto` here.

kdimak replied:
OK, we agreed with @Baha that can use `localkms.PubKeyBytesToHandle()` to convert public key bytes to key handle and use it with crypto. This will require some time to change, I will push the solution later on today.

kdimak replied:
I think we should also check a type of the crypto impl. It it's a `tinkcrypto`, which can convert public key bytes to `*keyset.Handle` and call crypto.Verify(). In contrast, if other crypto impl is used then it could not accept `*keyset.Handle`, in this case, we will get into trouble. What do you think? @Baha-sk  @troyronda

Baha-sk replied:
this is why `case *keyset.Handle:` will manage Tink wrapped keys, other implementations can support `case verify.PublicKey`

The API will be flexible with these two case conditions.

soluchok commented on pkg/doc/signature/ed25519signature2018/suite.go:
```diff
@@ -81,18 +93,28 @@ func (s *SignatureSuite) GetDigest(doc []byte) []byte {
+	switch pubKey := pubKeyValue.(type) {
+	case *verifier.PublicKey:
+		// ed25519 panics if key size is wrong
+		if l := len(pubKey.Value); l != ed25519.PublicKeySize {
```
`l := len(pubKey.Value); l != ed25519.PublicKeySize` -> `len(pubKey.Value) != ed25519.PublicKeySize`

kdimak commented on pkg/doc/signature/ed25519signature2018/suite_test.go:
```diff
@@ -42,37 +49,102 @@ func TestSignatureSuite_Sign(t *testing.T) {
+	}
+
+	err = ss.Verify(pubKey, doc, docSig)
+	require.NoError(t, err)
```
@Baha-sk the test is failing here saying that signature is invalid:
```
verify msg: verifier_factory: invalid signature
```
Do you have any idea why it's happening?

kdimak replied:
Filed an issue regarding this: https://github.com/hyperledger/aries-framework-go/issues/1483
I disabled accepting a public key bytes input when verifying a signature via `LocalCryptoVerifier`.

Baha-sk replied:
yeah, the signature has a special Tink prefix based on Tink key types, we can either fix it in Tink or manually remove the prefix from docSig and try `ss.Verify()` again... 

I will post an issue in Tink to see if we can get this added to Tink

kdimak commented:
I added `verifier` concept for the Signature Suite. It is defined by functional option in the same way as `signer` before. There are two default implementations of the `verifier`:
1.  `LocalCryptoVerifier` uses crypto and Local KMS for signature verification. The latter should be used to convert public key bytes to key handle and pass it to tinkcrypto (i removed support of this because of https://github.com/hyperledger/aries-framework-go/issues/1483).
2. `PublicKeyVerifier` is an old verification way using public key bytes. Other packages that use signature verification (like `did`  or `verifiable`) now use this type of verifier (mainly in the tests).

troyronda commented on pkg/doc/did/doc.go:
```diff
@@ -790,6 +795,16 @@ func (r *didKeyResolver) Resolve(id string) (*verifier.PublicKey, error) {
+func MapKeyTypeToKMS(t string) (kms.KeyType, error) {
+	switch t {
+	case "Ed25519VerificationKey2018":
+		return kms.Ed25519Type, nil
```
I'm a bit confused. What's the purpose of using interface{} above when we are mapping to definitions in kms anyways?

troyronda commented on pkg/doc/signature/ed25519signature2018/suite.go:
```diff
@@ -36,6 +42,89 @@ type signer interface {
+	var keyHandle *keyset.Handle
+
+	switch pubKey := pubKeyValue.(type) {
+	case *keyset.Handle:
```
I don't really understand this change. 
It seems like we are using an interface{} and then requiring deep knowledge of the types (coupling the kms types).

ed25519 package should not know about tink.

Baha-sk replied:
this is because `publicKeyValue` can be either a key reference (Tink's *keyset.Handle) or the raw []byte publicKey, similar to Crypto API that takes a `kh interface` argument for the key reference and the default implementation uses Tink's *keyset.Handle as kh type

troyronda commented on pkg/doc/signature/ed25519signature2018/suite.go:
```diff
@@ -36,6 +42,89 @@ type signer interface {
+
+// Verify will verify a signature.
+func (v *PublicKeyVerifier) Verify(pubKeyValue interface{}, doc, signature []byte) error {
+	pubKey, ok := pubKeyValue.(*sigverifier.PublicKey)
```
I'm a bit confused. It seems like we require a particular interface in this line, but not in the func signature?

Baha-sk replied:
`PublicKeyVerifier` is for keys in []byte format

troyronda commented on pkg/doc/signature/ed25519signature2018/suite.go:
```diff
@@ -36,6 +42,89 @@ type signer interface {
+		return errors.New("ed25519: bad public key length")
+	}
+
+	verified := ed25519.Verify(pubKey.Value, doc, signature)
```
this should be part of the kms/crypto implementation @Baha-sk

Baha-sk replied:
I think @kdimak wants to separate signature verification using raw key []byte from Tink keys and this verify uses []byte keys with crypto.ed25519 package.

So `PublicKeyVerifier` is for keys with raw bytes and `LocalKeyVerifier` is for Tink keys (with public key loaded from a DID doc as []byte and transformed by localkKMS into *keyset.Handle which requires a fix I'm currently working on)

troyronda replied:
This package should not know such details (that would be the reason for an empty interface).

troyronda commented on pkg/doc/signature/ed25519signature2018/suite.go:
```diff
@@ -36,6 +42,89 @@ type signer interface {
+
+// LocalCryptoVerifier defines a verifier which is based on Local KMS and Crypto
+// which uses keyset.Handle as input for verification.
+type LocalCryptoVerifier struct {
```
This doesn't belong in this package @Baha-sk @kdimak

Baha-sk replied:
Agree, this struct can do any signature verification, not just ed25519.

ed25519signaturesuite2018 can use this struct to do verification, as well as any NIST approved ecdsa signature suite.

troyronda replied:
More specifically, this package should not know what kms is in use nor anything about tink.

troyronda commented on pkg/doc/did/doc.go:
```diff
@@ -747,7 +747,7 @@ type signatureSuite interface {
 
 	// Verify will verify signature against public key
-	Verify(pubKey *verifier.PublicKey, doc []byte, signature []byte) error
+	Verify(pubKey interface{}, doc []byte, signature []byte) error
```
this change *only* makes sense if pubKey is an opaque object to every package except for a kms implementation.

troyronda replied:
and I also see resolve actually constructing a generic value... (so pubkey doesn't seem to be opaque at all).

troyronda commented on pkg/doc/did/doc.go:
```diff
@@ -777,11 +777,16 @@ type didKeyResolver struct {
 }
 
-func (r *didKeyResolver) Resolve(id string) (*verifier.PublicKey, error) {
+func (r *didKeyResolver) Resolve(id string) (interface{}, error) {
```
this change __only__ makes sense if pubKey is an opaque object to every package except for a kms implementation.

but I see this package constructing a generic return value. this is confusing.

sandrask replied:
I agree that we should not return interface here.

troyronda commented on pkg/doc/did/doc.go:
```diff
@@ -790,6 +795,16 @@ func (r *didKeyResolver) Resolve(id string) (*verifier.PublicKey, error) {
 }
 
+// MapKeyTypeToKMS maps key type of DID doc to the KMS one.
+func MapKeyTypeToKMS(t string) (kms.KeyType, error) {
```
This function doesn't belong in the did package.

kdimak replied:
Why? The key types are defined at https://www.w3.org/TR/did-core/#public-keys spec, and we need to map them to our internal key types.

kdimak commented:
Now there is no dependency on KMS/Crypto in `ed25519signature2018`.  `keyResolver` produces generic `interface{}` and passes it to the `verifier`. I left `PublicKeyVerifier` in the production code as it is still used in many places in aries. Also, crypto operations like `ed25519.Verify()` are still present as I don't see a way now of how to use Crypto directly for this purpose. E.g. I would like to have the possibility to call a Crypto function with the following definition: 
```
Verify(signature, msg []byte, kms.KeyType type, privKey []byte)
```
If present, I would call it `Verify(signature, msg, kms.Ed25519Type, privKey)` and that's it.
I also added naive implementation of `signer/verifier` based on `crypto.Crypto` in a separate file. In the tests, one can see how it can be implemented using tinkcrypto/localkms.

troyronda commented on pkg/doc/signature/ed25519signature2018/public_key_verifier.go:
```diff
@@ -0,0 +1,44 @@
+
+// Verify will verify a signature.
+func (v *PublicKeyVerifier) Verify(pubKeyValue interface{}, doc, signature []byte) error {
+	pubKey, ok := pubKeyValue.(*sigverifier.PublicKey)
```
This function should be calling a KMS implementation, not converting interfaces here... Is that the plan?

kdimak replied:
`NewCryptoVerifier()` is what we would like users to use. `PublicKeyVerifier` is a sort-of backward compatibility struct which is used in other packages (`verifiable`, `did`, `didexchange`,  `legacykms`) and should be replaced in the follow-up PRs. I can leave a comment regarding this in the code.

troyronda commented on pkg/doc/signature/ed25519signature2018/suite_crypto.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+// NewCryptoVerifier creates a new CryptoVerifier.
+func NewCryptoVerifier(cr crypto.Crypto) *CryptoVerifier {
```
Why is this only called by tests?

kdimak replied:
This will be changed in the follow-up PRs in order to not make the current PR that huge.

sandrask commented on pkg/doc/did/doc.go:
```diff
@@ -777,11 +777,16 @@ type didKeyResolver struct {
+func (r *didKeyResolver) Resolve(id string) (interface{}, error) {
 	for _, key := range r.PubKeys {
 		if key.ID == id {
+			keyType, err := MapKeyTypeToKMS(key.Type)
```
We could defer key type  mapping when it is needed (sign, verify) and not tie it to resolution. 
Did package should not know about KMS types. This should be postponed until sign or verify. You can have helper function on crypto before PubKeyBytesToHandle(pubKey.Value, pubKey.Type) is called.

kdimak replied:
I agree with you and @troyronda . I removed this conversion. I also changed the `Type` of the `PublicKey` to be just a string as we discussed on the call with @sandrask .

sandrask commented on pkg/doc/signature/ed25519signature2018/suite_crypto_test.go:
```diff
@@ -0,0 +1,110 @@
+	var keyHandle *keyset.Handle
+
+	switch pubKey := kh.(type) {
+	case *keyset.Handle:
```
I am not sure in what case handle would be passed in. Normally use case:
1. retrieve key from did (pub key bytes, type) 
2. use pub key for verification.

Baha-sk replied:
usually pub key bytes will be transformed into public keyset.Handle to use kms/crypto's default implementation services

kdimak replied:
I changed `publicKeyResolver.Resolve()` and `verifier.Verify()` to take `PublicKey` instead of genetic `interface{}`. The reason is that we don't see any use case to take e.g.  `*keyset.Handle` for now.
It looks like that when we would like to use `tinkcypto.Verifier`, we can intercept the call and convert `PublicKey` into the `*keyset.Handle` using `PubKeyBytesToHandle()`. This approach is implemented in **suite_crypto_test.go**.

troyronda commented on pkg/didcomm/protocol/didexchange/states.go:
```diff
@@ -591,10 +590,11 @@ func verifySignature(connSignature *ConnectionSignature, recipientKeys string) (
 
 	err = signatureSuite.Verify(&verifier.PublicKey{
-		Type:  kms.Ed25519Type,
+		Type:  "ED25519",
```
why this change?

sandrask replied:
I commented that we tie kms to our signature suite. Now that I see this it doesn't look good either. Dima please revert back. I expected this type to be passed in from resolver - not hard-coded.

kdimak replied:
The type is now a `string`, not `kms.KeyType`. I decided it will look simpler in the tests to use just constant instead of `string(kms.Ed25519Type)` (plus extra import).

kdimak replied:
I updated constants usage. As @troyronda suggested, I added string key type constants to the KMS and re-used it for KeyType constants.

sandrask replied:
sounds good

Baha-sk commented on pkg/doc/signature/ed25519signature2018/public_key_verifier.go:
```diff
@@ -0,0 +1,35 @@
+func (v *PublicKeyVerifier) Verify(pubKey *sigverifier.PublicKey, doc, signature []byte) error {
+	// ed25519 panics if key size is wrong
+	if len(pubKey.Value) != ed25519.PublicKeySize {
+		return errors.New("ed25519: bad public key length")
```
change "ed25519: bad public key length" to "ed25519: invalid key"

kdimak replied:
changed

Baha-sk commented on pkg/doc/signature/ed25519signature2018/public_key_verifier.go:
```diff
@@ -0,0 +1,35 @@
+
+	verified := ed25519.Verify(pubKey.Value, doc, signature)
+	if !verified {
+		return errors.New("signature doesn't match")
```
change "signature doesn't match" with "ed25519: invalid signature"

kdimak replied:
changed
